File: src/main/java/cn/taskeren/gtnn/common/ReversedRecipeRegistry.java
package cn.taskeren.gtnn.common;

import cn.taskeren.gtnn.machine.recipe.DisassemblerRecipes;
import gregtech.api.util.GTRecipe;
import gregtech.api.util.GTUtility;
import net.minecraft.item.ItemStack;

import java.util.ArrayList;
import java.util.List;

public class ReversedRecipeRegistry {

	private static final List<GTRecipe> REVERSED_RECIPES = new ArrayList<>(1024);

	public static void registerShapedRecipe(ItemStack output, Object[] recipe) {
		var reversed = GTUtility.reverseShapedRecipe(output, recipe);
		reversed.ifPresent(REVERSED_RECIPES::add);
	}

	public static void registerShapelessRecipe(ItemStack output, Object[] recipe) {
		var reversed = GTUtility.reverseShapelessRecipe(output, recipe);
		reversed.ifPresent(REVERSED_RECIPES::add);
	}

	public static void registerRecipesToDisassemblers() {
		REVERSED_RECIPES.forEach(DisassemblerRecipes::registerReversedCraftingRecipe);
	}

}

File: src/main/kotlin/cn/elytra/mod/gtnn/rewind/module/disassembler/ReversedRecipeRegistry.kt
package cn.elytra.mod.gtnn.rewind.module.disassembler

import cn.elytra.mod.gtnn.GtnhNoNerf
import cn.taskeren.gtnn.machine.recipe.DisassemblerRecipes
import gregtech.api.util.GTRecipe
import gregtech.api.util.GTUtility
import net.minecraft.item.ItemStack

object ReversedRecipeRegistry {

	/**
	 * The list used to store all the **REVERSED** recipes.
	 */
	internal val Registry = arrayListOf<GTRecipe>().apply { ensureCapacity(2048) }

	@JvmStatic
	fun registerShaped(output: ItemStack, recipe: Array<Any?>) {
		runCatching {
			GTUtility.reverseShapedRecipe(output, *recipe).ifPresent(Registry::add)
		}.onFailure {
			GtnhNoNerf.LOG.warn("An invalid reversed recipe was found with exception, output: {}, inputs: {}", output, recipe, it)
		}
	}

	@JvmStatic
	fun registerShapeless(output: ItemStack, recipe: Array<Any?>) {
		runCatching {
			GTUtility.reverseShapelessRecipe(output, *recipe).ifPresent(Registry::add)
		}.onFailure {
			GtnhNoNerf.LOG.warn("An invalid reversed recipe was found with exception, output: {}, inputs: {}", output, recipe, it)
		}
	}

	@JvmStatic
	internal fun registerAllReversedRecipes() {
		Registry.forEach { DisassemblerRecipes.registerReversedCraftingRecipe(it) }
	}

}
